amplipi/streams.py

```diff
@@ -1075,9 +1077,110 @@ def info(self) -> models.SourceInfo:
+    print(f'connecting {self.name} to {src}...')
+
+    if self.mock:
+      self._connect(src)
+      return
+
+    # Power on Bluetooth and enable discoverability
+    subprocess.run(args='bluetoothctl power on'.split(), preexec_fn=os.setpgrp)
+    subprocess.run(args='bluetoothctl discoverable on'.split(), preexec_fn=os.setpgrp)
+    subprocess.run(args='sudo btmgmt fast-conn on'.split(), preexec_fn=os.setpgrp)
+
+    # Start metadata watcher
+    src_config_folder = f"{utils.get_folder('config')}/srcs/{src}"
+    os.system('mkdir -p {}'.format(src_config_folder))
+    song_info_path = f'{src_config_folder}/currentSong'
+    device_info_path = f'{src_config_folder}/btDevice'
+    btmeta_args = f'{sys.executable} {utils.get_folder("streams")}/bluetooth.py --song-info={song_info_path} ' \
+                  f'--device-info={device_info_path} --output-device={utils.output_device(src)}'
+    self.bt_proc = subprocess.Popen(args=btmeta_args.split(), preexec_fn=os.setpgrp)
+
+    self._connect(src)
+    return
+
+  def _is_running(self):
+    if self.bt_proc:
+      return self.bt_proc.poll() is None
+    return False
+
+  def disconnect(self):
+    if self._is_running():
+      os.killpg(os.getpgid(self.bt_proc.pid), signal.SIGKILL)
+    self.bt_proc = None
+
+    # Power off Bluetooth and disable discoverability
+    subprocess.run(args='bluetoothctl discoverable off'.split(), preexec_fn=os.setpgrp)
+    subprocess.run(args='bluetoothctl power off'.split(), preexec_fn=os.setpgrp)
+
+    self._disconnect()
+
+  def info(self) -> models.SourceInfo:
+    src_config_folder = f"{utils.get_folder('config')}/srcs/{self.src}"
+    loc = f'{src_config_folder}/currentSong'
+    source = models.SourceInfo(name=self.full_name(),
+                               state=self.state,
+                               img_url=self.logo,
+                               supported_cmds=self.supported_cmds)
+    try:
+      with open(loc, 'r') as file:
+        data = json.loads(file.read())
+        source.artist = data['artist']
+        source.track = data['title']
+        source.album = data['album']
+        source.state = data['status']
+        return source
+    except Exception as e:
+      print(f'bluetooth: exception {e}')
+      traceback.print_exc()
+    return source
+
+  def send_cmd(self, cmd):
+    print(f'bluetooth: sending command {cmd}')
+    try:
+      if cmd in self.supported_cmds and self.src is not None:
+        src_config_folder = f"{utils.get_folder('config')}/srcs/{self.src}"
+        device_info_path = f'{src_config_folder}/btDevice'
+        btcmd_args = f'{sys.executable} {utils.get_folder("streams")}/bluetooth.py --command={cmd} --device-info={device_info_path}'
+        subprocess.run(args=btcmd_args.split(), preexec_fn=os.setpgrp)
+      else:
+        raise NotImplementedError(f'"{cmd}" is either incorrect or not currently supported')
+    except Exception as e:
+      print(f'bluetooth: exception {e}')
```

Comment:
We should probably re-rasie this: somethign like:
```python
      raise RuntimeError(f'Command {cmd} failed to send: {exc}') from e
```

Reply:
What is {exc} here?